src/gtpu/encap.c

```diff
@@ -830,18 +847,21 @@ static int gtp5g_fwd_skb_encap(struct sk_buff *skb, struct net_device *dev,
 
     pdr->ul_pkt_cnt++;
     pdr->ul_byte_cnt += skb->len; /* length without GTP header */
-    GTP5G_INF(NULL, "PDR (%u) UL_PKT_CNT (%llu) UL_BYTE_CNT (%llu)", pdr->id, pdr->ul_pkt_cnt, pdr->ul_byte_cnt);
-
+    GTP5G_INF(NULL, "PDR (%u) UL_PKT_CNT (%llu) UL_BYTE_CNT (%llu)", pdr->id, pdr->ul_pkt_cnt, pdr->ul_byte_cnt);    
+ 
+    if (pdr->urr_num != 0) {
+        if (check_urr(pdr, far, volume, volume_mbqe, true, drop_pkt) < 0)
+            GTP5G_ERR(pdr->dev, "Fail to send Usage Report");
+    }
+    
+    if (color == Red){
+        return -1;
+    }
     ret = netif_rx(skb);
     if (ret != NET_RX_SUCCESS) {
         GTP5G_ERR(dev, "Uplink: Packet got dropped\n");
```

Comment:
Should we return value other than 0?

Reply:
@chen042531 please reply.

Reply:
I changed this with macro COLOR_RED_DROP(0) not -1 for the reason that this is not error

Reply:
There are still mixed usages for 0, -1, COLOR_RED_DROP.
I think COLOR_RED_DROP should be value other than 0 (maybe -2?) so that we can distinguish different runtime behaviors (forwarded, dropped due to what failed reason, dropped due to QoS).

By the way, related error codes may need to be updated as well 
https://github.com/free5gc/gtp5g/blob/49fbe753afa79292ca0c09da5e640ed93dbac62d/src/gtpu/encap.c#L143

Reply:
I will update these part

Reply:
As discussed with @chen042531, only 1, 0, -1 should be returned from encap function. The return values will be replaced with macro, and only return value with -1 (dropped) should be freed.